framework/src/utils/MooseMeshXYCuttingUtils.C

```diff
@@ -0,0 +1,1230 @@
+        const auto node_m2 = mesh.add_point((*node_1 + *node_2 * 2.0) / 3.0);
+        makeImprovedTriElement(mesh,
+                               tri_group.first,
+                               ordered_node_list[side_counter],
+                               node_m1->id(),
+                               block_info.first,
+                               {boundary_to_improve},
+                               side_0_boundary_ids,
+                               std::vector<boundary_id_type>());
+        makeImprovedTriElement(mesh,
+                               tri_group.first,
+                               node_m1->id(),
+                               node_m2->id(),
+                               block_info.first,
+                               {boundary_to_improve},
+                               std::vector<boundary_id_type>(),
+                               std::vector<boundary_id_type>());
+        makeImprovedTriElement(mesh,
+                               tri_group.first,
+                               node_m2->id(),
+                               ordered_node_list[side_counter + block_info.second],
+                               block_info.first,
+                               {boundary_to_improve},
+                               std::vector<boundary_id_type>(),
+                               side_t_boundary_ids);
+      }
+      side_counter += block_info.second;
+    }
+    // TODO: Need to check if the new element is inverted?
+  }
+  // Delete the original elements
+  for (const auto & elem_to_remove : elems_to_remove)
+    mesh.delete_elem(mesh.elem_ptr(elem_to_remove));
+  mesh.contract();
+}
+
+void
+makeImprovedTriElement(ReplicatedMesh & mesh,
+                       const dof_id_type node_id_0,
+                       const dof_id_type node_id_1,
+                       const dof_id_type node_id_2,
+                       const subdomain_id_type subdomain_id,
+                       const std::vector<boundary_id_type> boundary_ids_for_side_1,
+                       const std::vector<boundary_id_type> boundary_ids_for_side_0,
+                       const std::vector<boundary_id_type> boundary_ids_for_side_2)
+{
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  Elem * elem_Tri3_new = mesh.add_elem(new Tri3);
+  elem_Tri3_new->set_node(0) = mesh.node_ptr(node_id_0);
+  elem_Tri3_new->set_node(1) = mesh.node_ptr(node_id_1);
+  elem_Tri3_new->set_node(2) = mesh.node_ptr(node_id_2);
+  for (const auto & boundary_id_for_side_0 : boundary_ids_for_side_0)
+    boundary_info.add_side(elem_Tri3_new, 0, boundary_id_for_side_0);
+  for (const auto & boundary_id_for_side_1 : boundary_ids_for_side_1)
+    boundary_info.add_side(elem_Tri3_new, 1, boundary_id_for_side_1);
+  for (const auto & boundary_id_for_side_2 : boundary_ids_for_side_2)
+    boundary_info.add_side(elem_Tri3_new, 2, boundary_id_for_side_2);
+  elem_Tri3_new->subdomain_id() = subdomain_id;
+}
```

Comment:
did you want to also preserve the extra element IDs?

Reply:
Okay, I allow merging of elements with the same subdomain ids here. They may have different extra element IDs though. Maybe I will just mention that improving TRI elements would lose the extra element IDs?

In the time, I will add extra element IDs preserve for the CUT_ELEM_TRI.

Reply:
Done. The elements with the same subdomain id but different extra element ids are not merged.